extending/plugins.rst

```diff
@@ -0,0 +1,134 @@
+            if "git clone" in conanfile_content and "git checkout" in conanfile_content:
+                fixed_sources = True
+                for invalid in invalid_content:
+                    if invalid in conanfile_content:
+                        fixed_sources = False
+            else:
+                fixed_sources = False
+                for valid in valid_content:
+                    if valid in conanfile_content:
+                        fixed_sources = True
+
+            if not fixed_sources:
+                output.error("%s Source files does not come from and immutable place. Checkout to a "
+                            "commit/tag or download a compressed source file" % test)
+
+This plugin is only checking the recipe content prior to the recipe being exported and prior to downloading the sources. Basically the
+``pre_export()`` function is checking the attributes of the ``conanfile`` object to see if there is an URL, a license and a description and
+warning the user with a message through the ``output``. This is done **before** the recipe is exported to the local cache.
+
+The ``pre_source()`` function checks if the recipe contains a ``source()`` method (this time it is using the conanfile content instead of
+the ``conanfile`` object) and in that case it checks if the download of the sources are likely coming from immutable places (a compressed
+file or a determined :command:`git checkout`). This is done **before** the **source()** method of the recipe is called.
+
+Any kind of Python scripting can be executed. You can create global functions and called them from different plugin functions, import them
+from a relative module and warn, error or even raise to block the Conan client execution.
+
+As you can see each function receives some parameters but not all of them are available for all functions as this may change depending on
+the context of the commands being executed such as the recipe being in the local cache or not.
+
+.. important::
+
+    A detailed description of the functions allowed and its parameters as well as their execution can be found in it dedicated reference
+    section: :ref:`plugins_reference`.
+
+Other useful task where a plugin may come handy are the upload and download actions. There are **pre** and **post** functions for every
+download/upload as a whole and for fine download task such as recipe and package downloads/uploads.
+
+For example they can be used to sign the packages before being uploaded and checking that signature when they are downloaded.
+
+.. code-block:: python
+   :caption: *signing_plugin.py*
+
+    import os
+    from conans import tools
+
+
+    def pre_upload_package(output, conanfile_path, reference, package_id, remote, **kwargs):
+        package_path = os.path.abspath(os.path.join(os.path.dirname(conanfile_path), "..", "package", package_id, "conan_package.tgz"))
+        my_signing_package_function(package_path)
```

Comment:
suggest modifying the package is not a good idea.

Reply:
Maybe we can live just with the `post_download_package` example